File: perceptron.js
function somatorio(entradas, pesos) {
  if (entradas.length != pesos.length) {
    console.log(" As entradas e pesos não possuem a mesma quantidade.")
    return
  }

  let soma = 0
  for (let i = 0; i < entradas.length; i++) {
    soma += parseFloat(entradas[i]) * parseFloat(pesos[i])
  }
  return soma
}

function comparador(limiteAtivacao, soma) {
  return soma >= limiteAtivacao ? 1 : 0
}

function corrigir(entrada, peso, taxaAprendizado, saidaDesejada, saidaAtual) {
  const deltaPeso = taxaAprendizado * (saidaDesejada - saidaAtual) * entrada
  const novoPeso = peso + deltaPeso

  console.log(`deltaPeso = ${deltaPeso}`)

  return parseFloat(novoPeso).toFixed(2)
}
let entradas = []
let pesos = []

entradas.push([3, 1, 2])
entradas.push([2, 7, 3])
entradas.push([3, 9, 3])
entradas.push([2, 2, 2])
entradas.push([9, 3, 9])
entradas.push([7, 2, 7])
entradas.push([1, 3, 1])
console.log(`Entradas = ${entradas.join(",")}`)

pesos.push([1])
pesos.push([1])
pesos.push([1])
pesos.push([1])
pesos.push([1])
pesos.push([1])
pesos.push([1])

console.log(`Pesos = ${pesos.join(",")}`)

const saida = [1, 0, 0, 0, 0, 0]
const saidaFinal = [0, 0, 0, 0, 0, 0]

const taxaAprendizado = 0.9
const limiteAtivacao = 0.8

while (true) {
  let flag = 1
  console.log({ pesos })

  for (let coluna = 0; coluna < entradas[0].length; coluna++) {
    let aux = []
    for (let linha = 0; linha < entradas.length; linha++) {
      aux.push(entradas[linha][coluna])
    }

    const soma = somatorio(aux, pesos)
    console.log(`Somatoria = ${soma}`)

    const saidaAtual = comparador(limiteAtivacao, soma)
    saidaFinal[coluna] = saidaAtual

    if (saidaAtual != saida[coluna]) {
      console.log("\nAjustando...")

      for (let i = 0; i < pesos.length; i++) {
        pesos[i] = parseFloat(
          corrigir(
            entradas[i][coluna],
            pesos[i],
            taxaAprendizado,
            saida[coluna],
            saidaAtual
          )
        )
        console.log(`\n peso ${i} = ${pesos[i]}`)
      }
      flag = 0
      break
    }
  }
  if (flag == 1) break
}

console.log("\nFINAL => N")
console.log(`Pesos = ${pesos.join(",")}`)
console.log(saidaFinal.join(","))

File: perceptron.py
def somatorio(xn, wn):
    if len(xn) != len(wn):
        print("A quantidade de entrada não bate")
        exit(0)
    soma = 0
    for i in range(len(xn)):
        soma += xn[i]*wn[i]
    return soma


def comparador(th, soma):
    if(soma >= th):  # th ?
        return 1
    else:
        return 0


def corrigir(x, w, n, t, o):
    # t => saida desejda
    # o => saida atual
    # n => taxa de aprendizado
    dw = n*(t-o)*x
    wi = w+dw  # ?
    print('dw =', dw)
    #print('w1 =',wi)
    return wi


x = []
x.append([3, 1, 2])
x.append([2, 7, 3])
x.append([3, 9, 3])
x.append([2, 2, 2])
x.append([9, 3, 9])
x.append([7, 2, 7])
x.append([1, 3, 1])
print('x =', x)


[
    [3, 1, 2],
    [2, 7, 3],
    [3, 9, 3],
    [2, 2, 2],
    [9, 3, 9],
    [7, 2, 7],
    [1, 3, 1],
]


w = []
w.append(1)
w.append(1)
w.append(1)
w.append(1)
w.append(1)
w.append(1)
w.append(1)
print('w =', w)

saida = [1, 0, 0, 0, 0, 0]

saidaFinal = [0, 0, 0, 0, 0, 0]

n = 0.9
th = 0.8

while(True):
    flag = 1
    print(w)
    for i in range(len(x[0])):
        aux = []
        for j in range(len(x)):
            aux.append(x[j][i])

        som = round(somatorio(aux, w), 2)
        print('Somatorio =', som)
        o = comparador(th, som)
        saidaFinal[i] = o
        if o != saida[i]:
            print("\nAjustando...")
            for j in range(len(w)):
                w[j] = round(corrigir(x[j][i], w[j], n, saida[i], o), 2)
                print('w'+str(j)+'=', w[j])
            flag = 0
            break
    if flag == 1:
        break

print("\nw =", w)
print(saidaFinal)
